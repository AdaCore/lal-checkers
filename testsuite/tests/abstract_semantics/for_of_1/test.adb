procedure Test is
   type Range_Type is range 0 .. 3;
   type Array_Type is array (Range_Type) of Integer;

   My_Array : Array_Type;
   X : Integer := 0;
begin
   My_Array(0) := 12;
   My_Array(1) := 4;
   My_Array(2) := 14;
   My_Array(3) := 7;

   for E of My_Array loop
      X := X + E;
   end loop;
end Ex1;
