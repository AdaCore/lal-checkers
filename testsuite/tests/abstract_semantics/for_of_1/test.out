{
  "always": {
    "assign0": [
      {
        "trace:": [
          "assign0", 
          "start0"
        ], 
        "values": {
          "E": "[-2147483648, 2147483647]", 
          "My_Array": "{([0, 3]): [-2147483648, 2147483647]}", 
          "X": "[-2147483648, 2147483647]"
        }
      }
    ], 
    "assign1": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "start0"
        ], 
        "values": {
          "E": "[-2147483648, 2147483647]", 
          "My_Array": "{([0, 3]): [-2147483648, 2147483647]}", 
          "X": "[0, 0]"
        }
      }
    ], 
    "assign2": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "start0"
        ], 
        "values": {
          "E": "[-2147483648, 2147483647]", 
          "My_Array": "{([0, 0]): [12, 12], ([1, 3]): [-2147483648, 2147483647]}", 
          "X": "[0, 0]"
        }
      }
    ], 
    "assign3": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "start0"
        ], 
        "values": {
          "E": "[-2147483648, 2147483647]", 
          "My_Array": "{([0, 0]): [12, 12], ([1, 1]): [4, 4], ([2, 3]): [-2147483648, 2147483647]}", 
          "X": "[0, 0]"
        }
      }
    ], 
    "assign4": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign4", 
          "start0"
        ], 
        "values": {
          "E": "[-2147483648, 2147483647]", 
          "My_Array": "{([0, 0]): [12, 12], ([1, 1]): [4, 4], ([2, 2]): [14, 14], ([3, 3]): [-2147483648, 2147483647]}", 
          "X": "[0, 0]"
        }
      }
    ], 
    "assign5": [
      {
        "trace:": [
          "assign0", 
          "assign1", 
          "assign2", 
          "assign3", 
          "assign4", 
          "assign5", 
          "start0"
        ],
        "values": {
          "E": "[-2147483648, 2147483647]",
          "My_Array": "{([0, 0]): [12, 12], ([1, 1]): [4, 4], ([2, 2]): [14, 14], ([3, 3]): [7, 7]}",
          "X": "[0, 0]"
        }
      }
    ],
    "assign6": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assign6",
          "start0"
        ],
        "values": {
          "E": "[-2147483648, 2147483647]",
          "My_Array": "{([0, 0]): [12, 12], ([1, 1]): [4, 4], ([2, 2]): [14, 14], ([3, 3]): [7, 7]}",
          "X": "[0, 0]"
        }
      }
    ],
    "assign7": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assign6",
          "assign7",
          "assume0",
          "loop_start0",
          "start0"
        ],
        "values": {
          "E": "[4, 14]",
          "My_Array": "{([0, 0]): [12, 12], ([1, 1]): [4, 4], ([2, 2]): [14, 14], ([3, 3]): [7, 7]}",
          "X": "[-2147483648, 2147483647]"
        }
      }
    ],
    "assume0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assign6",
          "assign7",
          "assume0",
          "loop_start0",
          "start0"
        ],
        "values": {
          "E": "[4, 14]",
          "My_Array": "{([0, 0]): [12, 12], ([1, 1]): [4, 4], ([2, 2]): [14, 14], ([3, 3]): [7, 7]}",
          "X": "[-2147483648, 2147483647]"
        }
      }
    ],
    "loop_join0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assign6",
          "assign7",
          "assume0",
          "loop_join0",
          "loop_start0",
          "start0"
        ],
        "values": {
          "E": "[-2147483648, 2147483647]",
          "My_Array": "{([0, 0]): [12, 12], ([1, 1]): [4, 4], ([2, 2]): [14, 14], ([3, 3]): [7, 7]}",
          "X": "[-2147483648, 2147483647]"
        }
      }
    ],
    "loop_start0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assign6",
          "assign7",
          "assume0",
          "loop_start0",
          "start0"
        ],
        "values": {
          "E": "[-2147483648, 2147483647]",
          "My_Array": "{([0, 0]): [12, 12], ([1, 1]): [4, 4], ([2, 2]): [14, 14], ([3, 3]): [7, 7]}",
          "X": "[-2147483648, 2147483647]"
        }
      }
    ],
    "start0": [
      {
        "trace:": [
          "start0"
        ],
        "values": {
          "E": "[-2147483648, 2147483647]",
          "My_Array": "{([0, 3]): [-2147483648, 2147483647]}",
          "X": "[-2147483648, 2147483647]"
        }
      }
    ]
  },
  "le_t_eq_v": {
    "assign0": [
      {
        "trace:": [
          "assign0",
          "start0"
        ],
        "values": {
          "E": "[-2147483648, 2147483647]",
          "My_Array": "{([0, 3]): [-2147483648, 2147483647]}",
          "X": "[-2147483648, 2147483647]"
        }
      }
    ],
    "assign1": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "start0"
        ],
        "values": {
          "E": "[-2147483648, 2147483647]",
          "My_Array": "{([0, 3]): [-2147483648, 2147483647]}",
          "X": "[0, 0]"
        }
      }
    ],
    "assign2": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "start0"
        ],
        "values": {
          "E": "[-2147483648, 2147483647]",
          "My_Array": "{([0, 0]): [12, 12], ([1, 3]): [-2147483648, 2147483647]}",
          "X": "[0, 0]"
        }
      }
    ],
    "assign3": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "start0"
        ],
        "values": {
          "E": "[-2147483648, 2147483647]",
          "My_Array": "{([0, 0]): [12, 12], ([1, 1]): [4, 4], ([2, 3]): [-2147483648, 2147483647]}",
          "X": "[0, 0]"
        }
      }
    ],
    "assign4": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "start0"
        ],
        "values": {
          "E": "[-2147483648, 2147483647]",
          "My_Array": "{([0, 0]): [12, 12], ([1, 1]): [4, 4], ([2, 2]): [14, 14], ([3, 3]): [-2147483648, 2147483647]}",
          "X": "[0, 0]"
        }
      }
    ],
    "assign5": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "start0"
        ],
        "values": {
          "E": "[-2147483648, 2147483647]",
          "My_Array": "{([0, 0]): [12, 12], ([1, 1]): [4, 4], ([2, 2]): [14, 14], ([3, 3]): [7, 7]}",
          "X": "[0, 0]"
        }
      }
    ],
    "assign6": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assign6",
          "start0"
        ],
        "values": {
          "E": "[-2147483648, 2147483647]",
          "My_Array": "{([0, 0]): [12, 12], ([1, 1]): [4, 4], ([2, 2]): [14, 14], ([3, 3]): [7, 7]}",
          "X": "[0, 0]"
        }
      }
    ],
    "assign7": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assign6",
          "assign7",
          "assume0",
          "loop_start0",
          "start0"
        ],
        "values": {
          "E": "[4, 14]",
          "My_Array": "{([0, 0]): [12, 12], ([1, 1]): [4, 4], ([2, 2]): [14, 14], ([3, 3]): [7, 7]}",
          "X": "[-2147483648, 2147483647]"
        }
      }
    ],
    "assume0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assign6",
          "assign7",
          "assume0",
          "loop_start0",
          "start0"
        ],
        "values": {
          "E": "[4, 14]",
          "My_Array": "{([0, 0]): [12, 12], ([1, 1]): [4, 4], ([2, 2]): [14, 14], ([3, 3]): [7, 7]}",
          "X": "[-2147483648, 2147483647]"
        }
      }
    ],
    "loop_join0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assign6",
          "assign7",
          "assume0",
          "loop_join0",
          "loop_start0",
          "start0"
        ],
        "values": {
          "E": "[-2147483648, 2147483647]",
          "My_Array": "{([0, 0]): [12, 12], ([1, 1]): [4, 4], ([2, 2]): [14, 14], ([3, 3]): [7, 7]}",
          "X": "[-2147483648, 2147483647]"
        }
      }
    ],
    "loop_start0": [
      {
        "trace:": [
          "assign0",
          "assign1",
          "assign2",
          "assign3",
          "assign4",
          "assign5",
          "assign6",
          "assign7",
          "assume0",
          "loop_start0",
          "start0"
        ],
        "values": {
          "E": "[-2147483648, 2147483647]",
          "My_Array": "{([0, 0]): [12, 12], ([1, 1]): [4, 4], ([2, 2]): [14, 14], ([3, 3]): [7, 7]}",
          "X": "[-2147483648, 2147483647]"
        }
      }
    ],
    "start0": [
      {
        "trace:": [
          "start0"
        ],
        "values": {
          "E": "[-2147483648, 2147483647]",
          "My_Array": "{([0, 3]): [-2147483648, 2147483647]}",
          "X": "[-2147483648, 2147483647]"
        }
      }
    ]
  }
}